### PR TITLE
Issue 20: Exercise Names + A Refactor to Bring Into Spec Compliance

### PR DESCRIPTION
**Pull request by BethanyG**
As stated on the 🥫 , a refactor to bring the analyzer "up to spec" 🌈  ✨  :

- [x] Altered (per #20 ) to read in/out/test from an exericses `config.json`
- [x] Altered to remove approve/disprove/refer to mentor top-level statuses
- [x]  Changed `Status` to `Summary` and added summary comments categories
- [x]  Added a `dataclass` to represent a `Comment`, with `CommentTypes`, `params` and `comments` text
- [x]  Added an `Enum` to represent `CommentTypes` for flagging each comment given.
- [x]  Mapped the feedback and message types given back from `pylint` to the `ComentTypes` defined for exercism. 
- [x]  Pulled `Pylint` specific code into its own file, since it was generic.
- [x]  Did some cleanup of names and imports
- [x]  Changed the output-format to conform to the spec listed here:  https://github.com/exercism/docs/blob/main/building/tooling/analyzers/interface.md#output-format
- [x]  Cleaned up tests to remove checks for the approve/disapprove/refer to mentor statuses
- [x]  Further refactored tests to account for `dataclass`, `Enum`, and `output_file` changes
- [x]  Pulled summary comment code out of `two-fer/analyzer`, made it generic and moved it to a `@classmethod` of `Analysis`

Also added CI-related files.

More work is needed to clean up code and better integrate `PyLint`, but this is a functional first pass.  🎉

### CODE REVIEW

**Review by SleeplessByte — APPROVED** [review]

File: dev-requirements.txt.py
Comment (BethanyG) [inline dev-requirements.txt.py:1]:
```suggestion
pytest>=5.3.1,!=5.4.0

```

File: lib/common/__init__.py
Comment (BethanyG) [inline lib/common/__init__.py:7]:
```suggestion
from .testing  import BaseExerciseTest

```

File: lib/common/comment.py
Comment (BethanyG) [inline lib/common/comment.py:71]:
```suggestion
        return f"{self.__class__.__name__}.{self.name}"
        
```

File: lib/common/pylint_comments.py
Comment (BethanyG) [inline lib/common/pylint_comments.py:39]:
```suggestion
    return pylint_comments
    
```

File: lib/common/analysis.py
Comment (iHiD) [inline lib/common/analysis.py:87]:
```suggestion
            print("I FOUND REQUIRED")
```

File: lib/common/comment.py
Comment (iHiD) [inline lib/common/comment.py:65]:
```suggestion
    GENERIC = "We don't have a custom analysis for this exercise yet, but here are some comments from PyLint to help you improve your code."
```

File: lib/common/comment.py
Comment (iHiD) [inline lib/common/comment.py:64]:
```suggestion
    INFORM = "Good work! Here are some general recommendations for improving your Python code."
```

File: lib/common/comment.py
Comment (iHiD) [inline lib/common/comment.py:61]:
```suggestion
    CELEBRATE = "Congratulations! This solution is very close to ideal! We don't have any specific recommendations."
```

File: test/two-fer/two_fer_test.py
Comment (iHiD) [inline test/two-fer/two_fer_test.py:29]: What's the advantage of adding these comments to all the test cases? I find it makes it a little harder to see the difference of each one.

**Review by iHiD — COMMENTED** [review]
I've made a few little comments.

One thing I'm not quite able to work out from a cursory glance is what happens if there are a mixture of comment types for one solution? If for example there is a `actionable` and a `informative` comment, what does the top-level message do?

File: test/two-fer/two_fer_test.py
Comment (SleeplessByte) [reply test/two-fer/two_fer_test.py:29]: (Don't want to speak for Bethany, but I followed the discussion a bit and I guess these docstrings are idiomatic, so adding them ensures the tests actually test the analyzer to ignore/find/parse them)

Comment (iHiD) [reply test/two-fer/two_fer_test.py:29]: Yeah - I can totally see one test that checks for docstrings. But presuming this gets some wider test coverage for a couple of dozen solutions it seems that adding them to all of them is going to add a lot of "noise" when trying to visually parse the tests?

Comment (cmccandless) [reply test/two-fer/two_fer_test.py:29]: > docstrings are idiomatic

Indeed they are

> adding them ensures the tests actually test the analyzer to ignore/find/parse them

This. Also, we have (or are going to have) a `docstrings` concept exercise, so the analyzer will need to be able to find/parse them

> it seems that adding them to all of them is going to add a lot of "noise" when trying to visually parse the tests?

I personally disagree; function and class docstrings are common enough in Python APIs that, to most Python developers, docstrings don't register as noise any more than `def` or `class` keywords; they are just a normal part of the syntax.

Comment (iHiD) [reply test/two-fer/two_fer_test.py:29]: OK. Seems strange to me. Having 7x more lines per example, with like 50 example, with no syntax highlighting to help you see through them, feels like a lot of extra noise for a brain to work through, especially in an analyzer where the differences between the test cases can be extremely subtle. But it's not my codebase so not something my opinion carries weight on. I'll mark this as resolved 🙂

**Comment by BethanyG** [comment]
> One thing I'm not quite able to work out from a cursory glance is what happens if there are a mixture of comment types for one solution? If for example there is a actionable and a informative comment, what does the top-level message do?

The "highest priority" comment takes precedence in that circumstance, and a a `DIRECT` (_"There are a few changes we suggest that can bring your solution closer to ideal."_ -- which maps to `ACTIONABLE` comments) summary message would be returned.  The logic is in a `@classmethod` on `Analysis` -- starting at **line 78** in `analysis.py`.  

Frankly, this was me not (quite) wanting to throw away the old APPROVE/DISAPPROVE organization - so I turned it into a summary.  Not sure I am happy with that, and I might get rid of it as I proceed down the road to integrating pylint better.   Open to thoughts/suggestions/ideas on that front.  

I did add a summary status of GENERIC, for when there were no exercism specific comments, and only pylint returned anything.  Haven't wired that up yet.  The comment Enums are in `comment.py`

One thought is to write a "generic" analyzer (with that GENERIC summary message) that ran a student's solution through pylint only, so that we could at minimum have a pylinting for every exercise.  Then double back and write/customize individual analyzers for exercises as we get to them.  Open to suggestions on that as well.   Pylint uses a pretty nice lib for AST wallking ([astriod](https://astroid.readthedocs.io/en/latest/)), that I'm eager to get my hands on.  😉

File: test/two-fer/two_fer_test.py
Comment (BethanyG) [reply test/two-fer/two_fer_test.py:29]: Chiming in late to the convo -- while DJ and Corey make some very good points about docstrings (_adding them can give you nice `help` messages and lightweight tests in the REPL for "free"_) --  and we do indeed intend to write one or more exercises about them -- 
the sad truth is that I added them into the tests to get `pylint` to **shut up already**.  😉    It kept failing even my CELEBRATORY code, which was causing bogus test case failure.  

I do find them very hard to read, especially with them defined as mulit-line constant strings in the test file.  So two things on that front:  
1. We may want to disable `pylint` warnings about docstrings on a per-exercise basis.  
2.  I am going to figure out some other method of making those test cases because that test file is making me really sad.  Current thought is that I would have a bunch of sample solutions in the test directory - which would also then test the reading/loading logic.  Something for the second refactor.  😄

Comment (iHiD) [reply test/two-fer/two_fer_test.py:29]: @BethanyG Cool 🙂 

- I'm quite pleased with the [Ruby approach](https://github.com/exercism/ruby-analyzer/blob/main/test/exercises/two_fer_test.rb) if you want some inspiration there: 
- @ErikSchierboom uses [fixtures instead in C#](https://github.com/exercism/csharp-analyzer/tree/main/test/Exercism.Analyzers.CSharp.IntegrationTests/Solutions), which is also nice

I also have disabled lots of rubocop checks for lots of this stuff as they're pretty unusual applications for it to work with.

**Comment by iHiD** [comment]
> Thank you for the comments (and the spelling patrol!!)! 😄

Of course 🙂 

> Open to thoughts/suggestions/ideas on that front.

I've not done this yet for the Ruby one, but my (medium-term) plan is to add summary comments where I feel they add some explicit extra value. The website itself should handle informing the student what to do in general (ie "please sort this before proceeding") so the aim of this field was for the analyzer to be able to give extra information where that wasn't enough. That said, in the short-term I'll probably do exactly the sort of thing you've done here. One to explore further down the line for sure 🙂

File: test/two-fer/two_fer_test.py
Comment (SleeplessByte) [reply test/two-fer/two_fer_test.py:29]: I have both inline solutions and fixture-files / directories where the latter have highlighting. Might be a cheap (*future PR) solution for the noise @BethanyG .

**Comment by BethanyG** [comment]
@cmccandless @iHiD @SleeplessByte  -- are there additional points I need to address here for approval?  Just checking....

**Comment by iHiD** [comment]
(Not from me, but you don't need my code-level-approval anyway as this doesn't touch anything I'm a codeowner of. I'll approve so you can merge when you're ready 🙂)

**Review by iHiD — APPROVED** [review]

File: dev-requirements.txt.py
Comment (cmccandless) [inline dev-requirements.txt.py:1]: I think this file is here by mistake?

**Review by cmccandless — APPROVED** [review]
I think dev-requirements.txt.py is an accident. Once that's removed, I'm good with this, so I'll approve now.

File: dev-requirements.txt.py
Comment (BethanyG) [reply dev-requirements.txt.py:1]: I put the file there to match the representer config here: https://github.com/exercism/python-representer.  The CI needs `pytest` for running tests, but the analyzer doesn't need `pytest` for its "normal" functioning.  Should I be adding `pytest` to the main `requirements.txt` instead?  What would be your suggested way of handling needing `pytest` for CI/dev testing?

Comment (BethanyG) [reply dev-requirements.txt.py:1]: UGH.  Thought I got rid of the `.py` version.  🤦🏽   Will delete.  Thanks!

**Comment by iHiD** [comment]
Whoop 🎉

**Pull request merged**
